### PR TITLE
Fix Kokkos::Vector::push_back for default execution space

### LINKED ISSUE #5042: Kokkos::vector::push_back() fails with certain Execution/Memory Space combination
Calling push_back() a second time zeros out the result of the first push_back().  There is a code example below that can be inserted into the Kokkos TestVector.hpp which demonstrates pushing onto an empty Kokkos::vector, checking the pushed value is correct, pushing a second value then checking that the first value is still correct which fails the check.  The first value gets reset to 0. 

The key is the memory space used.  The setting of the `TestGPUMemSpace` as shown in the code below is based on what the SPARC team has used for years as a way to make sure SPARC does not accidentally use UVM memory.  Note that the test runs on host, there are no kernel launches.   

This can be reproduced by inserting the code below into TestVector.hpp which causes the test to fail when run with 
   (1) nvidia/10.1.243
   (2) openmpi/4.0.3
   (3) gcc/7.2.0
This is the current SPARC development cuda-gcc combination.

Code Version:
  Kokkos Master : sha - b52f8c835f4df003954dad66d9761094f8baa66c

The platform is Skylake 4x18, dual V100  -  the ascicgpu16.sandia.gov platform.

cmake -D CMAKE_VERBOSE_MAKEFILE=FALSE -D CMAKE_BUILD_TYPE:STRING=RELEASE -D BUILD_SHARED_LIBS=OFF -D CMAKE_C_COMPILER=mpicc -D CMAKE_CXX_COMPILER=mpicxx -D CMAKE_CXX_FLAGS= -D CMAKE_EXE_LINKER_FLAGS= -D CMAKE_POSITION_INDEPENDENT_CODE=ON -D TPL_ENABLE_Pthread=OFF -D Kokkos_ENABLE_SERIAL=ON -D Kokkos_ENABLE_OPENMP=OFF -D Kokkos_ENABLE_PTHREAD=OFF -D TPL_ENABLE_CUDA=ON -D Kokkos_ENABLE_CUDA=ON -D Kokkos_ENABLE_CUDA_UVM=ON -D Kokkos_ARCH_VOLTA70=ON -D Kokkos_ENABLE_CUDA_LAMBDA=ON -D Kokkos_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE=ON -D KokkosContainers_ENABLE_TESTS:STRING=ON -D KokkosCore_ENABLE_TESTS:STRING=ON -D Kokkos_ENABLE_TESTS:STRING=ON -D Kokkos_ENABLE_DEBUG_BOUNDS_CHECK=OFF ..

6. code needed to reproduce the bug : listed below

8. command line needed to reproduce the bug: 
    KokkosContainers_UnitTest_Serial --gtest_filter=serial.vector_insert

8. please also attach the `KokkosCore_config.h` header file (generated during the build): below

**Any additional info**

Fails in DEBUG build also.

```
diff --git a/packages/kokkos/containers/unit_tests/TestVector.hpp b/packages/kokkos/containers/unit_tests/TestVector.hpp
index 57b92c3..48a1dc3 100644
--- a/packages/kokkos/containers/unit_tests/TestVector.hpp
+++ b/packages/kokkos/containers/unit_tests/TestVector.hpp
@@ -131,6 +131,59 @@ struct test_vector_insert {
     a.insert(a.end(), 2, 12);
 #endif
     a.insert(a.end(), b.begin(), b.end());
+
+#if defined(KOKKOS_ENABLE_CUDA) || defined(KOKKOS_ENABLE_HIP)
+  #if defined(KOKKOS_ENABLE_CUDA)
+    using TestGPUExecSpace       = Kokkos::Cuda;
+    using TestGPUMemSpace        = Kokkos::CudaSpace;
+  #elif defined(KOKKOS_ENABLE_HIP)
+    using TestGPUExecSpace       = Kokkos::Experimental::HIP;
+    using TestGPUMemSpace        = Kokkos::Experimental::HIPSpace;
+  #endif
+  typedef Kokkos::Device<TestGPUExecSpace::execution_space, TestGPUMemSpace::memory_space> TestVectorDeviceType;
+#else
+  typedef Kokkos::Device<Kokkos::DefaultExecutionSpace::execution_space, Kokkos::DefaultExecutionSpace::memory_space>
+    TestVectorDeviceType;
+#endif
+    typedef unsigned long long int UInt64;
+    typedef Kokkos::pair<int,UInt64> Pair;
+    typedef Kokkos::vector<Pair, TestVectorDeviceType> PairVec;
+    PairVec V;
+    V.clear();
+    {
+      const int first = 4;
+      const UInt64 second = 1;
+      const Pair p(first, second);
+      V.push_back(p);
+    }
+    if (V[0].first != 4 || V[0].second != 1)
+      std::cout<<__FILE__<<":"<<__LINE__
+        <<" ERROR during push_back expected V[0]:(4,1) found "
+        <<" V[0]:("<<V[0].first<<","<<V[0].second<<")"
+        <<" info: V.size():"<<V.size()<<" V.span:"<<V.span()<<std::endl;
+    ASSERT_EQ(V[0].first,4);
+    ASSERT_EQ(V[0].second,1u);
+
+    {
+      const int first = 3;
+      const UInt64 second = 2;
+      const Pair p(first, second);
+      V.push_back(p);
+    }
+    if (V[0].first != 4 || V[0].second != 1)
+      std::cout<<__FILE__<<":"<<__LINE__
+        <<" ERROR during push_back expected V[0]:(4,1) found "
+        <<" V[0]:("<<V[0].first<<","<<V[0].second<<")"
+        <<" info: V.size():"<<V.size()<<" V.span:"<<V.span()<<std::endl;
+    if (V[1].first != 3 || V[1].second != 2)
+      std::cout<<__FILE__<<":"<<__LINE__
+        <<" ERROR during push_back expected V[1]:(3,2) found "
+        <<" V[1]:("<<V[1].first<<","<<V[1].second<<")"
+        <<" info: V.size():"<<V.size()<<" V.span:"<<V.span()<<std::endl;
+    ASSERT_EQ(V[1].first,3);
+    ASSERT_EQ(V[1].second,2u);
+    ASSERT_EQ(V[0].first,4);
+    ASSERT_EQ(V[0].second,1u);
   }

   template <typename Vector>
```

KokkosCore_config.h as requested:

```

#if !defined(KOKKOS_MACROS_HPP) || defined(KOKKOS_CORE_CONFIG_H)
#error \
    "Do not include KokkosCore_config.h directly; include Kokkos_Macros.hpp instead."
#else
#define KOKKOS_CORE_CONFIG_H
#endif

// KOKKOS_VERSION % 100 is the patch level
// KOKKOS_VERSION / 100 % 100 is the minor version
// KOKKOS_VERSION / 10000 is the major version
#define KOKKOS_VERSION 30600

/* Execution Spaces */
#define KOKKOS_ENABLE_SERIAL
/* #undef KOKKOS_ENABLE_OPENMP */
/* #undef KOKKOS_ENABLE_OPENMPTARGET */
/* #undef KOKKOS_ENABLE_THREADS */
#define KOKKOS_ENABLE_CUDA
/* #undef KOKKOS_ENABLE_HIP */
/* #undef KOKKOS_ENABLE_HPX */
/* #undef KOKKOS_ENABLE_MEMKIND */
/* #undef KOKKOS_ENABLE_LIBRT */
/* #undef KOKKOS_ENABLE_SYCL */

#ifndef __CUDA_ARCH__
/* #undef KOKKOS_ENABLE_TM */
/* #undef KOKKOS_USE_ISA_X86_64 */
/* #undef KOKKOS_USE_ISA_KNC */
/* #undef KOKKOS_USE_ISA_POWERPCLE */
/* #undef KOKKOS_USE_ISA_POWERPCBE */
#endif

/* General Settings */
#define KOKKOS_ENABLE_CXX14
/* #undef KOKKOS_ENABLE_CXX17 */
/* #undef KOKKOS_ENABLE_CXX20 */

#define KOKKOS_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE
#define KOKKOS_ENABLE_CUDA_UVM
#define KOKKOS_ENABLE_CUDA_LAMBDA
/* #undef KOKKOS_ENABLE_CUDA_CONSTEXPR */
/* #undef KOKKOS_ENABLE_CUDA_LDG_INTRINSIC */
/* #undef KOKKOS_ENABLE_IMPL_CUDA_MALLOC_ASYNC */
/* #undef KOKKOS_ENABLE_HIP_RELOCATABLE_DEVICE_CODE */
/* #undef KOKKOS_ENABLE_HIP_MULTIPLE_KERNEL_INSTANTIATIONS */
/* #undef KOKKOS_ENABLE_HPX_ASYNC_DISPATCH */
/* #undef KOKKOS_ENABLE_DEBUG */
/* #undef KOKKOS_ENABLE_DEBUG_DUALVIEW_MODIFY_CHECK */
/* #undef KOKKOS_ENABLE_DEBUG_BOUNDS_CHECK */
/* #undef KOKKOS_ENABLE_COMPILER_WARNINGS */
/* #undef KOKKOS_ENABLE_PROFILING_LOAD_PRINT */
/* #undef KOKKOS_ENABLE_TUNING */
#define KOKKOS_ENABLE_DEPRECATED_CODE_3
#define KOKKOS_ENABLE_DEPRECATION_WARNINGS
/* #undef KOKKOS_ENABLE_LARGE_MEM_TESTS */
/* #undef KOKKOS_ENABLE_DUALVIEW_MODIFY_CHECK */
/* #undef KOKKOS_ENABLE_COMPLEX_ALIGN */
#define KOKKOS_ENABLE_IMPL_DESUL_ATOMICS
/* #undef KOKKOS_OPT_RANGE_AGGRESSIVE_VECTORIZATION */
/* #undef KOKKOS_ENABLE_AGGRESSIVE_VECTORIZATION */

/* TPL Settings */
/* #undef KOKKOS_ENABLE_HWLOC */
/* #undef KOKKOS_USE_LIBRT */
/* #undef KOKKOS_ENABLE_HBWSPACE */
#define KOKKOS_ENABLE_LIBDL
/* #undef KOKKOS_ENABLE_LIBQUADMATH */
/* #undef KOKKOS_IMPL_CUDA_CLANG_WORKAROUND */

#define KOKKOS_COMPILER_CUDA_VERSION 101

/* #undef KOKKOS_ARCH_SSE42 */
/* #undef KOKKOS_ARCH_ARMV80 */
/* #undef KOKKOS_ARCH_ARMV8_THUNDERX */
/* #undef KOKKOS_ARCH_ARMV81 */
/* #undef KOKKOS_ARCH_ARMV8_THUNDERX2 */
/* #undef KOKKOS_ARCH_AMD_AVX2 */
/* #undef KOKKOS_ARCH_AVX */
/* #undef KOKKOS_ARCH_AVX2 */
/* #undef KOKKOS_ARCH_AVX512XEON */
/* #undef KOKKOS_ARCH_KNC */
/* #undef KOKKOS_ARCH_AVX512MIC */
/* #undef KOKKOS_ARCH_POWER7 */
/* #undef KOKKOS_ARCH_POWER8 */
/* #undef KOKKOS_ARCH_POWER9 */
/* #undef KOKKOS_ARCH_INTEL_GEN */
/* #undef KOKKOS_ARCH_INTEL_DG1 */
/* #undef KOKKOS_ARCH_INTEL_GEN9 */
/* #undef KOKKOS_ARCH_INTEL_GEN11 */
/* #undef KOKKOS_ARCH_INTEL_GEN12LP */
/* #undef KOKKOS_ARCH_INTEL_XEHP */
/* #undef KOKKOS_ARCH_INTEL_GPU */
/* #undef KOKKOS_ARCH_KEPLER */
/* #undef KOKKOS_ARCH_KEPLER30 */
/* #undef KOKKOS_ARCH_KEPLER32 */
/* #undef KOKKOS_ARCH_KEPLER35 */
/* #undef KOKKOS_ARCH_KEPLER37 */
/* #undef KOKKOS_ARCH_MAXWELL */
/* #undef KOKKOS_ARCH_MAXWELL50 */
/* #undef KOKKOS_ARCH_MAXWELL52 */
/* #undef KOKKOS_ARCH_MAXWELL53 */
/* #undef KOKKOS_ARCH_PASCAL */
/* #undef KOKKOS_ARCH_PASCAL60 */
/* #undef KOKKOS_ARCH_PASCAL61 */
#define KOKKOS_ARCH_VOLTA
#define KOKKOS_ARCH_VOLTA70
/* #undef KOKKOS_ARCH_VOLTA72 */
/* #undef KOKKOS_ARCH_TURING75 */
/* #undef KOKKOS_ARCH_AMPERE */
/* #undef KOKKOS_ARCH_AMPERE80 */
/* #undef KOKKOS_ARCH_AMPERE86 */
/* #undef KOKKOS_ARCH_AMD_ZEN */
/* #undef KOKKOS_ARCH_AMD_ZEN2 */
/* #undef KOKKOS_ARCH_AMD_ZEN3 */
/* #undef KOKKOS_ARCH_VEGA */
/* #undef KOKKOS_ARCH_VEGA900 */
/* #undef KOKKOS_ARCH_VEGA906 */
/* #undef KOKKOS_ARCH_VEGA908 */
/* #undef KOKKOS_ARCH_VEGA90A */
```

### PR DESCRIPTION
**Pull request by masterleinad**
Fixes #5042. The main change is to allocate (and initialize) `create modified_flags` if it's not yet allocated in `modify`.
Without this, `resize` is copying the device data which has not been modified.

### CODE REVIEW

**Review by crtrott — APPROVED** [review]

File: containers/unit_tests/TestVector.hpp [hunk 1]
```diff
@@ -298,6 +298,36 @@ TEST(TEST_CATEGORY, vector_insert) {
   Impl::test_vector_insert<int, TEST_EXECSPACE>(3057);
 }
 
+TEST(TEST_CATEGORY, vector_push_back_default_exec) {
+  using TestVectorDeviceType =
+      Kokkos::Device<Kokkos::DefaultExecutionSpace,
+                     typename Kokkos::DefaultExecutionSpace::memory_space>;
```
Comment (dalg24) [inline containers/unit_tests/TestVector.hpp:304]:
```suggestion
                     Kokkos::DefaultExecutionSpace::memory_space>;
```

File: containers/unit_tests/TestVector.hpp [hunk 2]
```diff
@@ -298,6 +298,36 @@ TEST(TEST_CATEGORY, vector_insert) {
...
+  using UInt64  = unsigned long long int;
+  using Pair    = Kokkos::pair<int, UInt64>;
+  using PairVec = Kokkos::vector<Pair, TestVectorDeviceType>;
```
Comment (dalg24) [inline containers/unit_tests/TestVector.hpp:307]: The bug was not sensitive to the value type was it?  If not then prefer a scalar type.

File: containers/unit_tests/TestVector.hpp [hunk 3]
```diff
@@ -298,6 +298,36 @@ TEST(TEST_CATEGORY, vector_insert) {
   Impl::test_vector_insert<int, TEST_EXECSPACE>(3057);
 }
 
+TEST(TEST_CATEGORY, vector_push_back_default_exec) {
```
Comment (dalg24) [inline containers/unit_tests/TestVector.hpp:301]: Pick a better name or comment on why this test was added and what bug it exposes.  Optionally add a link to the issue.

**Review by dalg24 — APPROVED** [review]

**Comment by masterleinad** [comment]
@https://github.com/overfelt Can you confirm that this fixes your problem?

**Comment by overfelt** [comment]
Yes, I'll rerun my tests.  Takes a while to build.

**Comment by overfelt** [comment]
Yes, this fixes my problem.

**Comment by dalg24** [comment]
OpenMPTarget-ROCm-4.5 failure seems unrelated
```
lld: error: /tmp/libkokkoscore-gfx906-ade9a8-gfx906-290beb.o: symbol _HASHW_/var/jenkins/workspace/Kokkos@2/core/unit_test/tools/TestEventCorrectness_cpp__omp_outlined___22_wrapper has undefined version 2/core/unit_test/tools/TestEventCorrectness_cpp__omp_outlined___22_wrapper
lld: error: /tmp/libkokkoscore-gfx906-ade9a8-gfx906-290beb.o: symbol _HASHW_/var/jenkins/workspace/Kokkos@2/core/unit_test/tools/TestEventCorrectness_cpp__omp_outlined___24_wrapper has undefined version 2/core/unit_test/tools/TestEventCorrectness_cpp__omp_outlined___24_wrapper
clang-13: error: amdgcn-link command failed with exit code 1 (use -v to see invocation)
make[2]: *** [core/unit_test/CMakeFiles/KokkosCore_UnitTest_KokkosP.dir/build.make:136: core/unit_test/KokkosCore_UnitTest_KokkosP] Error 1
make[1]: *** [CMakeFiles/Makefile2:2294: core/unit_test/CMakeFiles/KokkosCore_UnitTest_KokkosP.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

**Review by nliber — APPROVED** [review]

**Pull request merged**